src/ui/marker.js

```diff
@@ -6,34 +6,46 @@ import LngLat from '../geo/lng_lat';
 import Point from '@mapbox/point-geometry';
 import smartWrap from '../util/smart_wrap';
 import { bindAll } from '../util/util';
+import { type Anchor, anchorTranslate, applyAnchorClass } from './anchor';
 
 import type Map from './map';
 import type Popup from './popup';
 import type {LngLatLike} from "../geo/lng_lat";
 import type {MapMouseEvent} from './events';
 
+type Options = {
+    offset?: PointLike,
+    anchor?: Anchor
+};
+
 /**
  * Creates a marker component
  * @param element DOM element to use as a marker. If left unspecified a default SVG will be created as the DOM element to use.
  * @param options
+ * @param options.anchor A string indicating the Marker's location relative to the coordinate set via {@link Marker#setLngLat}.
+ *   Options are `'middle'`, `'top'`, `'bottom'`, `'left'`, `'right'`, `'top-left'`, `'top-right'`, `'bottom-left'`, and `'bottom-right'`.
+ *   The default is `'middle'`.
  * @param options.offset The offset in pixels as a {@link PointLike} object to apply relative to the element's center. Negatives indicate left and up.
  * @example
  * var marker = new mapboxgl.Marker()
  *   .setLngLat([30.5, 50.5])
  *   .addTo(map);
  * @see [Add custom icons with Markers](https://www.mapbox.com/mapbox-gl-js/example/custom-marker-icons/)
  */
-class Marker {
+export default class Marker {
     _map: Map;
+    _anchor: Anchor;
     _offset: Point;
     _element: HTMLElement;
     _popup: ?Popup;
     _lngLat: LngLat;
     _pos: ?Point;
 
-    constructor(element: ?HTMLElement, options?: {offset: PointLike}) {
+    constructor(element: ?HTMLElement, options?: Options) {
         bindAll(['_update', '_onMapClick'], this);
 
+        this._anchor = options && options.anchor || 'middle';
```

Comment:
should we add a check here so that if a user provides something like `new mapboxgl.Marker({anchor: 'left'})` it is valid instead of having to use `new mapboxgl.Marker(null, {anchor: 'left'})`?

if we don't want the former to be valid, we should fire a better error because right now it is just `Uncaught TypeError: Cannot read property 'add' of undefined`

Reply:
Gonna send a separate PR for this in a sec.